OSDOCS-19870: adds main console docs RHCL (#112113)
openshift/openshift-docs: modules/proc-rhcl-publish-api-for-consumers.adoc

// Module included in the following assemblies:
//
// * develop/rhcl-api-management.adoc

:_mod-docs-content-type: PROCEDURE
[id="proc-rhcl-publish-api-for-consumers_{context}"]

> = Publish an API for other teams to discover and consume

Reviewer: I assume you've already started moving away from gerunds for your procedure titles here. That's okay with me, but be sure to be consistent.

+
[TIP]
====
Include the following documentation types so that API consumers can self-serve:

* API reference documentation
* Getting started guides
* Code examples or client SDKs
* Support or contact information
====
* *Lifecycle and Visibility*: Choose how you want to publish your API. *Published* makes the API product available for discovery by API consumers. *Draft* keeps the API product hidden.
//TODO add procedure to deprecate and retire APIs
* *API Key Approval*: Choose how to manage API key requests.

** *Automatic*: Grant API keys immediately when requested. Use this for internal APIs where you trust all consumers.
** *Manual*: Review each request before approval. Use this when you need to verify use cases or limit access to specific teams.
+
[NOTE]
====
The system automatically discovers your authentication method from the `AuthPolicy` CR:
+
* For API key authentication, API consumers request API keys through the web console.
* For OpenID Connect (OIDC) or JWT authentication, API consumers can look up your identity provider details and token endpoint.
====

. Click *Create*.

> . Optional. After you publish your API product, click your API product from the *Name* tab, select the *Overview* tab, then scroll to the bottom of the page and add your contact information by selecting the *pencil icon* in each field.

Reviewer: ```suggestion
. Optional. After you publish your API product, click your API product from the *Name* tab, select the *Overview* tab, scroll to the end of the page, and add your contact information by selecting the *pencil icon* in each field.
```

.Verification

. Confirm that consumers can discover and understand your API by using the web console:

> .. Navigate to *{prodname}* -> *API Products* and verify that your API product shows status *Published*.

Reviewer: ```suggestion
.. Navigate to *{prodname}* -> *API Products* and verify that your API product shows the *Published* status.
```

.. Ask a developer from another team to browse the catalog and verify the following information:
+
* Your API is displayed in the catalog.
* The description clearly explains what the API does.
* Documentation links work.
* Authentication requirements are visible.

.Next steps

* Developers can now discover your API in the catalog, request access, and receive credentials.

> * If you configured manual approval, review and approve access requests on the *API Key Approvals* page.

Reviewer: ```suggestion
* If you configured manual approval, you can review and approve access requests on the *API Key Approvals* page.
```

// Module included in the following assemblies:
//
// * develop/rhcl-api-management.adoc

:_mod-docs-content-type: PROCEDURE
[id="proc-rhcl-publish-api-for-consumers_{context}"]
= Publish an API for other teams to discover and consume

[role="_abstract"]
When you have an internal API that other teams need to integrate with, you can publish it in a self-service catalog so that developers can discover it, understand how to use it, and request access with the appropriate credentials. Use the {ocp} web console with {prodname}.

.Prerequisites

* You enabled the {prodname} {ocp} web console plugin.
* You have an `HTTPRoute` custom resource (CR) routing traffic to your API service.
* You configured an `AuthPolicy` CR to protect your API.
* You have the `api-owner` role bound in your namespace.
//Q: is this something we needed to do on the cluster itself?

.Procedure

. In the {ocp} web console, navigate to *{prodname}* -> *API Products*.

. Click *Create APIProduct*.

> . Describe your API for consumers by entering the following information:

Reviewer 2: We need to refine the available options here

<img width="993" height="1039" alt="Image" src="https://github.com/user-attachments/assets/ad95f222-a669-4b39-974c-248d496a8f5c" />

+

> * *Name*: Enter a unique resource name.

Reviewer 2: This is called "Kubernetes resource name"

Author: I removed these deep details

> * *Display Name*: Enter a clear name that describes what the API does (for example, "Payment Processing API" or "Customer Profile Service").

Reviewer 2: This is called "API product name"

Author: I'm just going to high-level the main areas and let the user fill in the fields (easier to maintain docs if we change any of these down the road)

* *Description*: Explain what problems the API solves and what use cases it supports. Help consumers understand whether this API meets their needs.

> 

Reviewer 2: Do we want to provide comprehensive list of fields?

If so, version, tags are missing

Author: Nope!

. Select the `HTTPRoute` CR that routes traffic to your API.

> . Configure how API consumers authenticate:

Reviewer 2: The owner does not configure how the API consumers authenticat. This given when the HTTPRoute is selected. The console plugin will report back to the user in the creation form the associated policies to the selected HTTPRoute. 

As enhancement, we should implement reporting the authentication method as well.

+
The system automatically discovers your authentication method from the `AuthPolicy` resource:
+
* For API key authentication, consumers request API keys through the console.
* For OIDC/JWT authentication, consumers see your identity provider details and token endpoint.

. Choose how to manage access requests:
+
* *Automatic*: Grant API keys immediately when requested. Use this for internal APIs where you trust all consumers.
* *Manual*: Review each request before approval. Use this when you need to verify use cases or limit access to specific teams.

> . Add the following documentation links so that API consumers can self-serve:

Reviewer 2: Missing the publish status: Published or Draft

Author: we should come back after the release and add retire and deprecate, too

+

> * API reference documentation

Reviewer 2: Only 
* _OpenAPI Spec URL  _: Enter the full path to your API spec file
Currently only by URL. I have also seen forms where your can paste embedded the content (yaml/json format) of the openapi doc. Should we add issue to implement it? @R-Lawton 

 * _Documentation URL _: Link to external documentation for this API

* Getting started guides
* Code examples or client SDKs
* Support or contact information

> 

Reviewer 2: Now I realize that the api product has a contact section
```yaml
    contact:
      email: api-owners@example.com
      slack: '#api-support'
      team: api-owners
```

That cannot be filled at creation time from the apiproduct creation form. Being optional, it will be created empty and then, on the overview tab of the api product you can fill the contact info. Should we add contact info to the creation form? @R-Lawton

Author: Added a step for now. LMK if it lands before release and I can remove it.

. Click *Create*.

> . Publish the API product to make it visible in the catalog by running the following command in your terminal window:

Reviewer 2: You can remove this patch. It's already in the creation form.

Author: awesome, that was my next question
